### PR TITLE
Adds possibility to define a goal description

### LINKED ISSUE #7346: Description/Comment on Goals
I experienced a certain problem about goals while working in my company. As some users from the management also have access to Piwik they try to analyze the traffic on our website. We have a lot of goals defined to measure the usage of our website. This goals only have relatively short names and use a certain naming convention so that we as the developers know in witch part of our web application this goals are placed. But this short names often aren't as clear to everybody as we thought and this leads to recurring misinterpretation of the goals. So it would be very helpful if there was the possibility to create a detailed description for goals. I guess this shouldn't be a lot of work, but it would be a huge improvement in usability and collaborative usage of Piwik.

### PR DESCRIPTION
**Pull request by sgiehl**
replaces #9469 

image

image

fixes #7346

### CODE REVIEW

**Comment by sgiehl** [comment]
@tsteur In the old PR I had changed the goal page to show the description as kind of subheadline under the goal title or as tooltip. See:

image

image

Is there any simple possibility to do that with the new container/widget framework?

**Comment by tsteur** [comment]
Not super easily unless it was a feature of the HtmlTable or EvolutionGraph visualization but I presume it should appear with any visualization?

If it should work with any visualization there are multiple ways to solve it. We could eg enrich `WidgetConfig` if it should be a feature of all widgets https://github.com/piwik/piwik/blob/3.x-dev/core/Widget/WidgetConfig.php or we could add a `$description` property to https://github.com/piwik/piwik/blob/3.x-dev/core/Report/ReportWidgetConfig.php which would be only for Report widgets. I think `ReportWidgetConfig` would not be right though and it should maybe rather add a new `description` property to `API.getReportMetadata` via the `Report` class https://github.com/piwik/piwik/blob/3.x-dev/core/Plugin/Report.php#L516 . In JS we already merge `WidgetMetadata` and `ReportMetadata` to get the report documentation see https://github.com/piwik/piwik/blob/3.x-dev/plugins/CoreHome/angularjs/widget/widget.directive.js#L61-L67 . Here we could then also merge a `widget.description = report.description` and display it conditionally below the headline (if present) https://github.com/piwik/piwik/blob/3.x-dev/plugins/CoreHome/angularjs/widget/widget.directive.html#L12 . It will then work for all reports. Is this kind of understandable?

* Add a description to reportMetadata
* Merge description in widget.directive.js
* Show it in widget.directive.html 

Maybe there would be a better name than `description`? Not sure if it is confusing to have `documentation` vs `description`.

**Comment by sgiehl** [comment]
Thx for the explanation. Need to have a closer look at that tomorrow.

**Comment by tsteur** [comment]
I would recommend to not work on this for now. Looks like we will move the rendering of the headline back into the widget etc. Will keep you updated.

**Comment by tsteur** [comment]
The headline for reports and everything is now rendered in `plugins/CoreHome/templates/_dataTable.twig`. To make it possible to show the description we would possibly need to add a new ViewDataTable config property in `core/ViewDataTable/Config.php` and then check if a value is set in the twig template and if so, print it :) That should work, let me know if not

**Comment by sgiehl** [comment]
I've rebased the branch and updated the new angular javascript. Will have a look at the rendering of the description the next days...

**Comment by mattab** [comment]
LGTM. Minor feedback:

 * Maybe the input field could be wider (but still on one line) as description will be longer than the name, when set. 
 * an integration test is failing `test_getGoal_shouldReturnExistingGoal`
 * Left couple comments

**Comment by sgiehl** [comment]
I've updated the code, so the description will now be shown in the reports. failing tests should also be fixed

File: plugins/Goals/API.php [hunk 1]
```diff
@@ -219,6 +225,11 @@ private function checkName($name)
         return urldecode($name);
     }
 
+    private function checkDescription($description)
+    {
+        return urldecode($description);
```
Comment (mattab) [inline plugins/Goals/API.php]: as the description field is VARCHAR 255 maybe we should check here that the description is maximum 255 chars?

**Review by mattab — COMMENTED** [review]
The text is displayed in Black below the Goal name, whereas it was in grey in your original screenshot. was it intentional to show it in black?

**Comment by mattab** [comment]
Cheers @sgiehl - I'm merging already but feel free to make minor changes post-merge if needed

File: plugins/Goals/API.php (hunk 1 above)
Comment (sgiehl) [inline plugins/Goals/API.php]: name is varchar 50 and not checked re length as well. Shall we add such checks?

Comment (mattab) [inline plugins/Goals/API.php]: good point. It's not needed then!

**Pull request merged**